pkg/util/log/gen.go

```diff
@@ -0,0 +1,429 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+// +build ignore
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"text/template"
+
+	"github.com/cockroachdb/cockroach/pkg/cli/exit"
+	"github.com/cockroachdb/errors"
+	"github.com/cockroachdb/gostdlib/go/format"
+)
+
+func main() {
+	if err := run(); err != nil {
+		fmt.Fprintln(os.Stderr, "ERROR:", err)
+		exit.WithCode(exit.UnspecifiedError())
+	}
+}
+
+func run() error {
+	if len(os.Args) < 3 {
+		return errors.Newf("usage: %s <proto> template>\n", os.Args[0])
+	}
+
+	// Which template are we running?
+	tmplName := os.Args[2]
+	tmplSrc, ok := templates[tmplName]
+	if !ok {
+		return errors.Newf("unknown template: %q", tmplName)
+	}
+	tmpl, err := template.New(tmplName).Parse(tmplSrc)
+	if err != nil {
+		return errors.Wrap(err, tmplName)
+	}
+
+	// Read the input .proto file.
+	chans, sevs, err := readInput(os.Args[1])
+	if err != nil {
+		return err
+	}
+
+	// Render the template.
+	var src bytes.Buffer
+	if err := tmpl.Execute(&src, struct {
+		Severities []info
+		Channels   []info
+	}{sevs, chans}); err != nil {
+		return err
+	}
+
+	// If we are generating a .go file, do a pass of gofmt.
+	newBytes := src.Bytes()
+	if strings.HasSuffix(tmplName, ".go") {
+		newBytes, err = format.Source(newBytes)
+		if err != nil {
+			return errors.Wrap(err, "gofmt")
+		}
+	}
+
+	// Write the output file.
+	w := os.Stdout
+	if len(os.Args) > 3 {
+		f, err := os.OpenFile(os.Args[3], os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0666)
+		if err != nil {
+			return err
+		}
+		defer func() { _ = f.Close() }()
+		w = f
+	}
+	if _, err := w.Write(newBytes); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type info struct {
+	RawComment string
+	Comment    string
+	PComment   string
+	Name       string
+	NAME       string
```

Comment:
why not `NameUpper`? 

I think you cal also use `call strings.Lower "bob"` in a template.